releases-implementation/README.md

```diff
@@ -0,0 +1,415 @@
+- All apps, both those deployed as App resources and as HelmRelease resources, MUST be added to Release `.spec.apps`. For every app:
+	- we MUST specify app name,
+	- we MUST specify app version,
+	- we CAN OPTIONALLY specify the name of the catalog from which the app is installed (“default” by default), and
+	- we CAN OPTIONALLY specify the app’s dependencies.
+- Kubernetes version MUST BE specified as a `.spec.components` entry,
+- Flatcar version and image variant MUST BE specified as `.spec.components` entries,
+- cluster-$provider app version MUST BE specified as a `.spec.components` entry.
+
+With the above in mind, a Release resource that would correspond to the current cluster-aws release would like like this:
+
+```yaml
+apiVersion: release.giantswarm.io/v1alpha1
+kind: Release
+metadata:
+  name: v25.0.0
+spec:
+  apps:
+  - name: aws-ebs-csi-driver
+    version: 2.30.1
+    dependsOn:
+    - cloud-provider-aws
+  - name: aws-pod-identity-webhook
+    version: 1.14.2
+    dependsOn:
+    - cert-manager
+  - name: capi-node-labeler
+    version: 0.5.0
+  - name: cert-exporter
+    version: 2.9.0
+    dependsOn:
+    - kyverno
+  - name: cert-manager
+    version: 3.7.5
+    dependsOn:
+    - prometheus-operator-crd
+  - name: chart-operator-extensions
+    version: 1.1.2
+    dependsOn:
+    - prometheus-operator-crd
+  - name: cilium
+    version: 0.24.0
+  - name: cilium-crossplane-resources
+    version: 0.1.0
+  - name: cilium-servicemonitors
+    version: 0.1.2
+    dependsOn:
+    - prometheus-operator-crd
+  - name: cloud-provider-aws
+    version: 1.25.14-gs2
+    dependsOn:
+    - vertical-pod-autoscaler-crd
+  - name: cluster-autoscaler
+    version: 1.27.3-gs8
+    dependsOn:
+    - kyverno
+  - name: coredns
+    version: 1.21.0
+  - name: external-dns
+    version: 3.1.0
+    dependsOn:
+    - prometheus-operator-crd
+  - name: metrics-server
+    version: 2.4.2
+    dependsOn:
+    - kyverno
+  - name: net-exporter
+    version: 1.19.0
+    dependsOn:
+    - prometheus-operator-crd
+  - name: network-policies
+    version: 0.1.0
+    catalog: cluster
+  - name: node-exporter
+    version: 1.19.0
+    dependsOn:
+    - kyverno
+  - name: vertical-pod-autoscaler
+    version: 5.1.0
+    dependsOn:
+    - prometheus-operator-crd
+  - name: vertical-pod-autoscaler-crd
+    version: 3.0.0
+  - name: etcd-k8s-res-count-exporter
+    version: 1.10.0
+    dependsOn:
+    - kyverno
+  - name: observability-bundle
+    version: 1.3.4
+    dependsOn:
+    - coredns
+  - name: k8s-dns-node-cache
+    version: 2.6.2
+    dependsOn:
+    - kyverno
+  - name: security-bundle
+    version: 1.6.5
+    catalog: giantswarm
+    dependsOn:
+    - prometheus-operator-crd
+  - name: teleport-kube-agent
+    version: 0.9.0
+  components:
+  - name: cluster-aws
+    version: 1.0.0
+  - name: flatcar
+    version: 3815.2.2
+  - name: flatcar-variant
+    version: 3.0.0
+  - name: kubernetes
+    version: 1.25.16
```

Comment:
so this would mean for an OS or K8s upgrade we would not need to release any new app/chart, but just bump the version here (and have the image available), right?

Reply:
I'm not sure that'd work technically. _Something_ would need to trigger the chart to lookup the new versions and change the image reference in the rendered templates.

Reply:
I just meant that with this proposal we would not need to release a new chart in many cases, only changes that need new values being wired up would need a new chart.

Reply:
> so this would mean for an OS or K8s upgrade we would not need to release any new app/chart

Yes, that is exactly the idea, when you need new OS/k8s/app version, you don't have to touch cluster-$provider app repo, no new versions are needed.

Then only scenario where cluster-$provider app changes are needed would be if some new k8s/OS/app version requires some changes in some config, but this is usually rare (and probably means it is some breaking change). We can see how to improve this scenario later, e.g. if needed we could decouple even app config from the cluster-$provider app and take it out, but I don't think this is needed (at least not now).

> but just bump the version here (and have the image available), right?

Since releases are immutable, we would create a new Release with the new versions. Then we would bump the release version for the cluster.

Reply:
sure, I did mean bump and release as new version, not bump inside existing version